openstack/baremetal/v1/nodes/requests.go

```diff
@@ -0,0 +1,274 @@
+	DeployWait   ProvisionState = "wait call-back"
+	Deploying    ProvisionState = "deploying"
+	DeployFail   ProvisionState = "deploy failed"
+	DeployDone   ProvisionState = "deploy complete"
+	Deleting     ProvisionState = "deleting"
+	Deleted      ProvisionState = "deleted"
+	Cleaning     ProvisionState = "cleaning"
+	CleanWait    ProvisionState = "clean wait"
+	CleanFail    ProvisionState = "clean failed"
+	Error        ProvisionState = "error"
+	Rebuild      ProvisionState = "rebuild"
+	Inpsecting   ProvisionState = "inspecting"
+	InspectFail  ProvisionState = "inspect failed"
+	InspectWait  ProvisionState = "inspect wait"
+	Adopting     ProvisionState = "adopting"
+	AdoptFail    ProvisionState = "adopt failed"
+	Rescue       ProvisionState = "rescue"
+	RescueFail   ProvisionState = "rescue failed"
+	Rescuing     ProvisionState = "rescuing"
+	UnrescueFail ProvisionState = "unrescue failed"
+)
+
+// ListOpts allows the filtering and sorting of paginated collections through
+// the API. Filtering is achieved by passing in struct field values that map to
+// the node attributes you want to see returned. Marker and Limit are used
+// for pagination.
+type ListOpts struct {
+	// Filter the list by specific instance UUID
+	InstanceUUID string `q:"instance_uuid"`
+
+	// Filter the list by chassis UUID
+	ChassisUUID string `q:"chassis_uuid"`
+
+	// Filter the list by maintenance set to True or False
+	Maintenance bool `q:"maintenance"`
+
+	// Nodes which are, or are not, associated with an instance_uuid.
+	Associated bool `q:"associated"`
+
+	// Only return those with the specified provision_state.
+	ProvisionState ProvisionState `q:"provision_state"`
+
+	// Filter the list with the specified driver.
+	Driver string `q:"driver"`
+
+	// Filter the list with the specified resource class.
+	ResourceClass string `q:"resource_class"`
+
+	// Filter the list with the specified conductor_group.
+	ConductorGroup string `q:"conductor_group"`
+
+	// Filter the list with the specified fault.
+	Fault string `q:"fault"`
+
+	// One or more fields to be returned in the response.
+	Fields []string `q:"fields"`
+
+	// Requests a page size of items.
+	Limit int `q:"limit"`
+
+	// The ID of the last-seen item.
+	Marker string `q:"marker"`
+
+	// Sorts the response by the requested sort direction.
+	SortDir string `q:"sort_dir"`
+
+	// Sorts the response by the this attribute value.
+	SortKey string `q:"sort_key"`
+
+	// Whether to show detailed information about the resource.
+	Detail bool `q:"detail"`
+
+	// A string or UUID of the tenant who owns the baremetal node.
+	Owner string `q:"owner"`
+}
+
+// ToNodeListQuery formats a ListOpts into a query string.
+func (opts ListOpts) ToNodeListQuery() (string, error) {
+	q, err := gophercloud.BuildQueryString(opts)
+	return q.String(), err
+}
+
+// List makes a request against the API to list nodes accessible to you.
+func List(client *gophercloud.ServiceClient, opts ListOptsBuilder) pagination.Pager {
```

Comment:
Something I don't quite get here is whether you handle detailed/non-detailed queries. I.e. in Ironic (cargo culted from Nova) we have /v1/nodes returning a list of summaries, /v1/nodes/detail returning a list of full nodes. In some API version we also introduced /v1/nodes?detail=True. Is it handled here?

Reply:
To get the list via `/v1/nodes?detail=True`:

```go
nodes.List(client, nodes.ListOpts{
    Detail: true,
})
```
However, the [API docs say `/v1/nodes/detail` is deprecated](https://developer.openstack.org/api-ref/baremetal/?expanded=list-nodes-detailed-detail) so I did not implement it - should I?

Reply:
It depends on which versions of Ironic you target. detail=True is only available since Rocky, which is quite recent. I guess to support older versions we still need /v1/nodes/detail, maybe with fallback to /v1/nodes?detail=True if microversion is large enough.

Reply:
This is most likely relevant to this discussion: https://github.com/gophercloud/gophercloud/issues/1440

Reply:
Sort of, except `/v1/nodes/detail` is deprecated, the modern way works as in the example code above, because the correct URL is `/v1/nodes?detail=True`. 

What to do about older Ironic is getting into microversions discussion that I would rather not try to solve.

I believe the correct way to do this is have my code be aware of the microversion, and do the right thing for older Ironics if Detail is set in ListOpts, by hitting the different URL.  That requires comparison of microversion, which I think is probably would need pulling in something like https://github.com/hashicorp/go-version to correctly compare versions.

Reply:
If you'd prefer to avoid diving into microversions here and now, you need support for the deprecated endpoint, since it works for all existing microversions.

Reply:
There is no code in gophercloud today to handling specific microversion cases which require comparison, and because this library supports many other things other than just Ironic I think it's a better thing to happen in a dedicated PR as I may not consider all the issues here for everyone. 

For the detail endpoint specifically, I don't like how it looks with two different ways to do things with no hints to the user which they should use. One way to avoid that is I *could* remove the Detail option from ListOpts, and have everyone go through the ListDetail function for a detailed listing.  Once the above concerns about microversions are addressed, add the microversion support later to hit the new or old URLs, but for now hit `/v1/nodes/detail` for everyone.  What I want to avoid is having two ways to show listing with details.  Does that make sense?

Reply:
Yes, I agree.

Reply:
> I believe the correct way to do this is have my code be aware of the microversion,

Your application code should be aware of microversions.

> For the detail endpoint specifically, I don't like how it looks with two different ways to do things with no hints to the user which they should use.

The user should interact with an application which interacts with Gophercloud. The application should contain the logic which determines how to use Gophercloud.

Think of Gophercloud as one step above a generic REST library.

Reply:
Sorry - user in this context was developer interacting with gophercloud directly. I've pushed some changes that I think address the concerns here.